src/transformers/trainer.py

```diff
@@ -887,21 +907,70 @@ def _objective(trial):
             best_trial = study.best_trial
             best_run = BestRun(str(best_trial.number), best_trial.value, best_trial.params)
         elif self.hp_search_backend == HPSearchBackend.RAY:
-            # The TensorBoard writer does not pickle so we have to remove it (if it exists) while doing the ray hp
-            # search.
+            # The model and TensorBoard writer do not pickle so we have to remove them (if they exists)
+            # while doing the ray hp search.
             _tb_writer = self.tb_writer
             self.tb_writer = None
+            _model = self.model
+            self.model = None
             # Setup default `resources_per_trial` and `reporter`.
             if "resources_per_trial" not in kwargs and self.args.n_gpu > 0:
-                kwargs["resources_per_trial"] = {"gpu": self.args.n_gpu}
+                # `args.n_gpu` is considered the total number of GPUs that will be split
```

Comment:
This is all starting to be a bit too long for Trainer. Is it possible to factor this in a helper function that would lie in integrations?